esmvalcore/preprocessor/_time.py

```diff
@@ -201,44 +168,36 @@ def _extract_datetime(cube, start_datetime, end_datetime):
     time_coord = cube.coord('time')
     time_units = time_coord.units
     if time_units.calendar == '360_day':
-        if start_datetime.day > 30:
-            start_datetime = start_datetime.replace(day=30)
-        if end_datetime.day > 30:
-            end_datetime = end_datetime.replace(day=30)
-
-    t_1 = PartialDateTime(year=int(start_datetime.year),
-                          month=int(start_datetime.month),
-                          day=int(start_datetime.day),
-                          hour=int(start_datetime.hour),
-                          minute=int(start_datetime.minute),
-                          second=int(start_datetime.second))
-
-    t_2 = PartialDateTime(year=int(end_datetime.year),
-                          month=int(end_datetime.month),
-                          day=int(end_datetime.day),
-                          hour=int(end_datetime.hour),
-                          minute=int(end_datetime.minute),
-                          second=int(end_datetime.second))
-
-    constraint = iris.Constraint(time=lambda t: t_1 <= t.point < t_2)
-    cube_slice = cube.extract(constraint)
+        if isinstance(start_datetime.day, int) and start_datetime.day > 30:
+            start_datetime.day = 30
+        if isinstance(end_datetime.day, int) and end_datetime.day > 30:
+            end_datetime.day = 30
+
+    if not cube.coord_dims(time_coord):
+        constraint = iris.Constraint(
+            time=lambda t: start_datetime <= t.point < end_datetime)
+        cube_slice = cube.extract(constraint)
+    else:
+        # Convert all time points to dates at once, this is much faster
+        # than using a constraint.
+        dates = time_coord.units.num2date(time_coord.points)
+        select = (dates >= start_datetime) & (dates < end_datetime)
+        if select.any():
+            cube_slice = cube.subset(time_coord[select])
+        else:
+            cube_slice = None
+
     if cube_slice is None:
+        def dt2str(time: PartialDateTime) -> str:
+            txt = f"{time.year}-{time.month:02d}-{time.day:02d}"
+            if any([time.hour, time.minute, time.second]):
+                txt += f" {time.hour:02d}:{time.minute:02d}:{time.second:02d}"
+            return txt
         raise ValueError(
-            f"Time slice {start_datetime.strftime('%Y-%m-%d')} "
-            f"to {end_datetime.strftime('%Y-%m-%d')} is outside "
+            f"Time slice {dt2str(start_datetime)} "
+            f"to {dt2str(end_datetime)} is outside "
             f"cube time bounds {time_coord.cell(0)} to {time_coord.cell(-1)}.")
```

Comment:
I wonder if this can be improved even further, right now it looks something like this:

```bash
ValueError: Time slice 1860-01-01 to 1860-01-02 is outside cube time bounds Cell(point=cftime.DatetimeNoLeap(1850, 1, 16, 12, 0, 0, 0, has_year_zero=True), bound=(cftime.DatetimeNoLeap(1850, 1, 1, 0, 0, 0, 0, has_year_zero=True), cftime.DatetimeNoLeap(1850, 2, 1, 0, 0, 0, 0, has_year_zero=True))) to Cell(point=cftime.DatetimeNoLeap(1850, 1, 16, 12, 0, 0, 0, has_year_zero=True), bound=(cftime.DatetimeNoLeap(1850, 1, 1, 0, 0, 0, 0, has_year_zero=True), cftime.DatetimeNoLeap(1850, 2, 1, 0, 0, 0, 0, has_year_zero=True))).
```

Reply:
looks OK to me, but don't mind me, I am more comfortable with cryptic terminal-based stuff than with actual windows and GUI stuff, but I see what you mean :grin:  :+1:

Reply:
Fixed now. The error message looked fine if the time coordinate had no bounds (as was the case in the tests), but looked horrible if there were bounds.